Faster Draw Functions

Kethku opened the pull request:
Rewrite of the CanvasApi functions to use a frame buffer instead of the default javascript 2d canvas draw calls.

I tried to simplify things where I could. Since we no longer use the default draw calls, we cannot depend on the context object to keep track of the camera position. So I rewrote most of the draw functions to reuse the setPixel function which keeps track of the camera position instead.

I also changed the structure of the index.js file in canvasAPI/ so that draw functions can depend on each other. I think this makes things easier to understand and keeps things DRY but I can change it back if that would be preferable.

Most of the cassettes work fine as far as I can tell except for the boot animation which seems to have draw problems. At this point it is too late to figure out what is really going on, so I am going to head to bed and will look into it further.

TODO:
- Fix Boot Animation
- Deal with the Actor Magic in Iframe.js
- Verify that camera functions work as expected
- Do some more general testing

Kethku commented on src/iframe/src/canvasAPI/line.js:
```diff
@@ -1,11 +1,10 @@
     ;[x1, y1] = [y1, x1]
     ;[x2, y2] = [y2, x2]
-    steep = true
+    steep = true;
```
remove semi colon

Kethku commented on src/iframe/src/canvasAPI/rect.js:
```diff
@@ -0,0 +1,21 @@
+export const drawRectStroke = ({ x, y, w, h, c, line }) => {
+  let left = x;
```
Remove semicolons

Kethku commented on src/iframe/src/canvasAPI/sprite.js:
```diff
@@ -1,45 +1,19 @@
-  sprites,
-  ctx
-}) => {
+const drawSprite = ({ x, y, spriteIndex, darken, flipH, flipV, setPixel, sprites }) => {
```
Remove semicolons

Kethku commented on src/iframe/src/css/Iframe.css:
```diff
@@ -189,12 +189,12 @@ input {
 }
 ul {
-  display: flex;
+  display: grid;
```
Revert this :/ Or at least double check if it is supposed to be changed.

gabrielflorit replied:
You can revert this, I unversioned `Iframe.css`. It's a build artifact. My mistake for versioning it.

gabrielflorit commented:
Thank you for this. Taking a look now. One thing I forgot to mention: I'm using Prettier with these two options:

```{
  "singleQuote": true,
  "semi": false
}
```

Would you be able to format your commits likewise?

Kethku commented:
Of course, sorry I think I added comments for most of the semi colons, but I forgot the quotation marks. My bad 😲

gabrielflorit commented:
Ah no worries. Prettier is great. It fixes your code, so I rarely think about typing it the right way. It'll change your code to suit its defaults.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -575,6 +585,7 @@ class Iframe extends Component {
 
         // Draw this state.
         window.draw && window.draw(this.store.getState())
+        this.drawPixels()
```
I wonder if there's a way we can create a wrapper draw function so we avoid calling `drawPixels` after every `window.draw`?

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -839,6 +852,7 @@ class Iframe extends Component {
                   selectedActors.includes(d.name)
                 )
               })
+              this.drawPixels()
```
Sorry for my ignorance - why is this call needed?

Kethku replied:
Its not. My bad. I added it while trying to fix drawActors, but this will get fixed later.

gabrielflorit commented on src/iframe/src/canvasAPI/print.js:
```diff
@@ -1,70 +1,26 @@
+  let currentX = Math.floor(x - cameraX)
+  let currentY = Math.floor(y - cameraY)
+
+  for (let letter of letters.toString().split('')) {
```
In general, would you be able to add comments wherever possible? I try to do it every line, or close to every line. I don't trust I will remember what I did more than 2 weeks ago. ;)

Kethku replied:
Of course. I will comment this stuff a bunch more tonight.

gabrielflorit commented:
This is a really big commit. I wonder if we could have both rendering setups side-by-side? i.e. don't get rid of the old setup, add the new pixel buffer system. Then we could try it for a while before turning off the old one.

gabrielflorit commented:
We could specify the renderer in the query string, e.g. `https://script-8.github.io/?id=1fa7a9d8ec2d63b61ac47c8f5d2de162&renderer=framebuffer`, and then `Iframe.js` would use that instead.

Kethku commented:
I worry that the cost of maintaining both simultaneously is somewhat big.

Kethku commented:
Of course I'm biased :P

gabrielflorit commented:
I don't think we'd have both for a long period of time. Under a month. I definitely wouldn't want both! That would be a self-inflicted wound.

Kethku commented:
Fair enough. I will add back the original canvasAPI and move the new version to newCanvasAPI. I will then check the url to pick which one in the drawUserGraphics call.

Kethku commented:
I would also add though that the only functions that got big changes were print, map and the rect ones. The others are straight forward ports removing the single pixel rect call in favor of setPixel.

Print I re-implemented whole sale because it was spending a lot of time trying to optimize performance concerns that aren't an issue with the frame buffer. It was also doing camera transforms manually which is not necessary when setPixel is used. I will add a ton more comments to describe the actual algorithm but its pretty simple.

Map did similar tricks I assume for performance reasons. I changed it to just reuse the sprite command instead of drawing each sprite manually.

Rect was a simple pass through to the canvas context. I will add comments.

The rest are almost identical ports replacing draw commands with the equivalent setPixel, or line calls.

I do understand though that its a big change, and clearly there are still some bugs. So if you want to maintain both, I don't mind. I'm just not sure its too bad especially since the biggest cassettes all work identically.

gabrielflorit commented:
Sounds good, thank you for the explanation.

Kethku commented:
Ok at this point I have fixed every graphical bug that I know about.

1. I rewrote the circle algorithm to more exactly match the original results as well as be a little simpler
2. I commented the print algorithm like you requested
3. I removed semi colons and quotations where I found them. Hopefully I didn't miss any :)
4. I added back the old api and added a check for a search query to pick which to use.

Unfortunately I couldn't figure out how to keep the URL search query from changing. When I add the renderer param, the url just switches back to the url without the param. So if you could help me finish up that last part, that would be great.

Other than that, I think this PR is getting pretty close to checkin ready.

Kethku commented:
Oh duh. I still have to fix the actor magic.  I will have to think some more about how to fix that properly... Any ideas or suggestions would be appreciated.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -348,8 +359,18 @@ class Iframe extends Component {
+  }
+
   componentDidMount() {
-    this.updateGlobals()
```
I think this needs to stay here. It sets all the default globals. It's confusing and I should rename it.

Kethku replied:
I moved this update globals to the bottom of componentDidMount because it now needs to happen after the renderer property gets set. I looked at the code between and I don't think it depends on the globals, so I figured moving it later is not a terrible strategy.

An alternative option would be to move the URLSearchParams call to the top of the function and set the renderer there before calling updateGlobals. Up to you.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -478,6 +499,9 @@ class Iframe extends Component {
     }
+
+    this.renderer = params.get('renderer') || "framebuffer"
+    this.updateGlobals()
```
Why do we need to call `this.updateGlobals()` here?

Kethku replied:
See above

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -574,7 +598,7 @@ class Iframe extends Component {
 
         // Draw this state.
-        window.draw && window.draw(this.store.getState())
+        this.drawUserGraphics(this.store.getState());
```
semicolons! i'd recommend you run prettier on everything, then you can totally forget about it

Kethku replied:
Finally got it running. I run windows and prettier has some extra dependencies on windows.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -930,7 +954,11 @@ class Iframe extends Component {
             height={CANVAS_SIZE}
             ref={_canvas => {
-              this._canvas = _canvas
+              if (_canvas) {
```
I'm not sure we should have anything in here besides setting the react ref. I get a feeling it's an antipattern. But not sure.

Kethku replied:
Fixed.

gabrielflorit commented on src/iframe/src/frameBufferCanvasAPI/circle.js:
```diff
@@ -0,0 +1,62 @@
+const drawCircle = ({ cx, cy, radius, color, onlyStroke, setPixel, line }) => {
```
ooh this is all exciting to read!

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -173,6 +174,10 @@ class Iframe extends Component {
       sound: true
     }
+
+    this._pixelBuffer = new ArrayBuffer(4 * CANVAS_SIZE * CANVAS_SIZE)
```
For my sake more than anything else, could you add a comment on each of these three lines, explaining what they do? It will make it easier for me to understand the rest of the new shiny renderer.

Kethku replied:
Done.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -314,21 +319,27 @@ class Iframe extends Component {
     let globals
 
     if (!providedGlobals) {
+      let canvasAPI = contextCanvasAPI
```
We should use a ternary here `const canvasAPI = this.renderer === 'framebuffer' ? frameBufferCanvasAPI : contextCanvasAPI`

Kethku replied:
Okeydokey.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -314,21 +319,27 @@ class Iframe extends Component {
           sprites: this.state.sprites,
           map: this.state.map
         }),
+        log: this.logger,
```
Probably best to move this back to where it was originally, less noise

Kethku replied:
Fixed.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -478,6 +499,9 @@ class Iframe extends Component {
       window.addEventListener('message', handleMessage)
     }
+
+    this.renderer = params.get('renderer') || "framebuffer"
```
I think we should be clearer about what `this.renderer` means. Right now I feel like it's the actual renderer API. We should probably make it `this.useFrameBufferRenderer`.

Kethku replied:
Fixed.

gabrielflorit commented on src/iframe/src/colors.js:
```diff
@@ -21,6 +21,20 @@ const triplets = [
   [8, 20, 30]
 ]
 
+let intLookup = [];
```
Will this logic be run every time the `colors.js` file is read? Should it be part of a function instead?

gabrielflorit replied:
Also I think this whole file needs a bit of documentation. At the very least, I think we should have an example of inputs and outputs on function headers. But I can do that later.

Kethku replied:
Nah ecmascript modules (what webpack emulates in bundled files) only initialize once and the reference to the module is passed every time it is imported.

This does mean that the lookup will be populated at startup if some other bundle imports the color file, but its not a lot of computation for the benefit in runtime.

gabrielflorit commented on src/iframe/src/colors.js:
```diff
@@ -36,12 +50,17 @@ const colors = {
+
   // NOTE: if triplet isn't a color in the pallet, this will return undefined.
-  lookup (triplet) {
+  lookupTriplet (triplet) {
```
yes, better name

gabrielflorit commented:
OK I took another closer look. This is getting close to merge ready indeed. As soon as you finish up the small things I noticed, I will merge. Then we can work on:

- actors
- query param

Thanks again for making it more separate. This makes it easier to understand.

Kethku commented:
Okeydokey. I think I have responded to all of your comments.

gabrielflorit commented on src/iframe/src/Iframe.js:
```diff
@@ -173,6 +174,27 @@ class Iframe extends Component {
+    // Uint8ClampedArray and Uint32Array. Both the TypedArrays below refer to
+    // the same backing buffer, so modifying values via one will be reflected in
+    // the other.
+    this._pixelBuffer = new ArrayBuffer(4 * CANVAS_SIZE * CANVAS_SIZE)
```
I had no idea this is how it worked. Thank you for the explanation.
